# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [1]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass  \
797          798         1       3   
627          628         1       1   
168          169         0       1   
381          382         1       3   
427          428         1       2   
310          311         1       1   
889          890         1       1   
547          548         1       2   
221          222         0       2   
154          155         0       3   

                                                  Name     Sex   Age  SibSp  \
797                                   Osman, Mrs. Mara  female  31.0      0   
627                      Longley, Miss. Gretchen Fiske  female  21.0      0   
168                                Baumann, Mr. John D    male   NaN      0   
381                        Nakid, Miss. Maria ("Mary")  female   1.0      0   
427  Phillips, Miss. Kate Florence ("Mrs Kate Louis...  female  19.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
547                         Padro y Manent, Mr. Julian    male   NaN      0   
221                               Bracken, Mr. James H    male  27.0      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   

     Parch         Ticket     Fare Cabin Embarked  
797      0         349244   8.6833   NaN        S  
627      0          13502  77.9583    D9        S  
168      0       PC 17318  25.9250   NaN        S  
381      2           2653  15.7417   NaN        C  
427      0         250655  26.0000   NaN        S  
310      0          11767  83.1583   C54        C  
889      0         111369  30.0000  C148        C  
547      0  SC/PARIS 2146  13.8625   NaN        C  
221      0         220367  13.0000   NaN        S  
154      0      Fa 265302   7.3125   NaN        S

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [2]:
passengers['Survived'].sample(10)

99     0
723    0
469    1
570    1
456    0
438    0
653    1
282    0
434    0
727    1
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [3]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [4]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [5]:
# todo write code here
passengers['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [6]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [7]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [8]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [9]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [10]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass                                Name  \
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
381          382         1       3         Nakid, Miss. Maria ("Mary")   
651          652         1       2                 Doling, Miss. Elsie   
547          548         1       2          Padro y Manent, Mr. Julian   
144          145         0       2          Andrew, Mr. Edgardo Samuel   

        Sex   Age  SibSp  Parch         Ticket      Fare        Cabin  \
679    male  36.0      0      1       PC 17755  512.3292  B51 B53 B55   
381  female   1.0      0      2           2653   15.7417          NaN   
651  female  18.0      0      1         231919   23.0000          NaN   
547    male   NaN      0      0  SC/PARIS 2146   13.8625          NaN   
144    male  18.0      0      0         231945   11.5000          NaN   

    Embarked AgeCat  
679        C  Adult  
381        C  Child  
651        S  Child  
547        C    NaN  
144        S  Child

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [11]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [12]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [13]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [14]:
#sex_all_series = passengers['Sex'].value_counts()
#sex_survivor_series = survivors['Sex'].value_counts()

#sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
#sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
#sex_comparision_df

#

age_survivors = passengers[passengers['Survived'] ==1]
age_survivors['PassengerId'].count()

age_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe
agecat_comparison_df = pd.DataFrame({ 'AllPassengers' : age_all_series, 'Survivors' : agecat_survivor_series })
agecat_comparison_df['AgeCatSurvivalRate'] = agecat_comparison_df['Survivors'] / agecat_comparison_df['AllPassengers']
agecat_comparison_df

AllPassengers  Survivors  AgeCatSurvivalRate
Adult            575        220            0.382609
Child            139         70            0.503597

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [25]:
#age_all_series = passengers['AgeCat'].value_counts()
#agecat_survivor_series = survivors['AgeCat'].value_counts()

#agecat_comparison_df = pd.DataFrame({ 'AllPassengers' : age_all_series, 'Survivors' : agecat_survivor_series })
#agecat_comparison_df['AgeCatSurvivalRate'] = agecat_comparison_df['Survivors'] / agecat_comparison_df['AllPassengers']
#agecat_comparison_df

# todo: repeat the analysis in the previous cell for Pclass
passengers['Pclass'].value_counts()
passengers['Pclass'].value_counts(normalize=True)

Pclass_series = passengers['Pclass'].value_counts()
Pclass_survivor_series = survivors['Pclass'].value_counts()

Pclass_series_df = pd.DataFrame({ 'AllPassengers' : Pclass_series, 'Pclass' : Pclass_survivor_series})
Pclass_series_df['PclassSurvivalRate'] = Pclass_series_df['Pclass'] / Pclass_series_df['AllPassengers']
Pclass_series_df


AllPassengers  Pclass  PclassSurvivalRate
1            216     136            0.629630
2            184      87            0.472826
3            491     119            0.242363